lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,712 @@
+# [FinetuningScheduler](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html#finetuning_scheduler.fts.FinetuningScheduler) uses [FTSCheckpoint](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts_supporters.html#finetuning_scheduler.fts_supporters.FTSCheckpoint) (an extension of ``ModelCheckpoint``) to maintain schedule state with special metadata.
+#
+#
+# ```python
+# from pytorch_lightning import Trainer
+# from finetuning_scheduler import FinetuningScheduler
+# trainer = Trainer(callbacks=[FinetuningScheduler()])
+# trainer.fit(..., ckpt_path="some/path/to/my_checkpoint.ckpt")
+# ```
+#
+# Training will resume at the depth/level of the provided checkpoint according the specified schedule. Schedules can be altered between training sessions but schedule compatibility is left to the user for maximal flexibility. If executing a user-defined schedule, typically the same schedule should be provided for the original and resumed training sessions.
+#
+# By default ([FinetuningScheduler.restore_best](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html?highlight=restore_best#finetuning_scheduler.fts.FinetuningScheduler.params.restore_best) is ``True``), [FinetuningScheduler](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html#finetuning_scheduler.fts.FinetuningScheduler) will attempt to restore the best available checkpoint before finetuning depth transitions.
+#
+# ```python
+# trainer = Trainer(callbacks=[FinetuningScheduler()])
+# trainer.fit(..., ckpt_path="some/path/to/my_kth_best_checkpoint.ckpt")
+# ```
+#
+# Note that similar to the behavior of [ModelCheckpoint](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html), (specifically [this PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/12045)),
+# when resuming training with a different [FTSCheckpoint](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts_supporters.html#finetuning_scheduler.fts_supporters.FTSCheckpoint) ``dirpath`` from the provided
+# checkpoint, the new training session's checkpoint state will be re-initialized at the resumption depth with the provided checkpoint being set as the best checkpoint.
+
+# %% [markdown]
+# <div class="alert alert-warning">
+#
+# **Note:** Currently, [FinetuningScheduler](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html#finetuning_scheduler.fts.FinetuningScheduler) supports the following ``StrategyType``s:
+#
+# - ``DP``
+# - ``DDP``
+# - ``DDP_SPAWN``
+# - ``DDP_SHARDED``
+# - ``DDP_SHARDED_SPAWN``
```

Comment:
qq: any specific reason why this doesn't work with other strategies?

Reply:
Great question! For context, I verify with a per-commit multi-gpu pipeline (using an Azure self-hosted runner) that the Finetuning Scheduler extension works well with all of the strategies currently supported by PyTorch Lightning Lite except for `deepspeed`.
 
Unfortunately, DeepSpeed does not currently have a method available in its API to add/distribute optimizer parameter groups to an existing optimizer. Because of the way it wraps and then initializes the optimizer (flattening, partitioning and distributing the parameters etc..), I don't think DeepSpeed will work effectively with Finetuning Scheduler until a hook to add optimizer parameter groups to an existing DeepSpeed wrapped optimizer is made available. I'm definitely keeping an eye on the DeepSpeed API occasionally to see if such functionality is added and will gladly enable Finetuning Scheduler support if it is. If I get the bandwidth and a little more ambitious, I know I could try to submit a PR to them but I've got other priorities and ambitions at the moment. :smile: 
 
I'm similarly monitoring the FSDP API as well and will add support if a hook to add optimizer pgs is added to the API (or if I get more bandwidth try to add it myself). FSDP actually has some helpful optimizer methods already that I think would make adding this functionality myself easier than with DeepSpeed but I have to prioritize my work right now.
 
I haven't used Horovod much honestly but given that it wraps and distributes optimizers I suspected it would take a fair amount of work to enable Finetuning Scheduler compatibility to have deprioritized that for now.
 
The Bagua strategy was not available when I submitted this PR but I've taken a brief and it looks awesome! It very well may work and I'm planning to explore support for it for the next Finetuning Scheduler minor release (0.2, paired with PyTorch Lightning 1.7).
 
Finally, I'm pretty sure TPU_SPAWN will work with Finetuning Scheduler, but am somewhat hesitant to support it since I don't have an automated testing pipeline (like you guys do with a circleci GCP workflow for TPUs) to ensure extensive regular testing. I know I could probably hack a manual colab-driven TPU_SPAWN testing flow but hopefully it's okay with you if we go ahead with publishing the tutorial using Finetuning Scheduler 0.1.x in the meantime and potentially update Finetuning Scheduler to support Bagua (not relevant to the tutorial per se) and potentially TPU_SPAWN (which could be used in the tutorial) in the future.
 
WHEW! Maybe more context than you wanted but hopefully helpful! :smile: